lib/src/facts/aix/processor_resolver.cc

```diff
@@ -0,0 +1,141 @@
+#include <internal/facts/aix/processor_resolver.hpp>
+#include <leatherman/logging/logging.hpp>
+#include <leatherman/util/scope_exit.hpp>
+
+#include <cstdint>
+#include <stdexcept>
+#include <odmi.h>
+#include <sys/cfgodm.h>
+
+using namespace std;
+
+// unlike just about every other error code to string function in the
+// world, odm_err_msg doesn't just return its pointer - it wants a
+// char** argument to put it in. This wraps that ugly API in a more
+// strerror()-like interface for ease of use.
+static const char* odm_error_string() {
+    static char* msg;
+    int result = odm_err_msg(odmerrno, &msg);
+    if (result < 0) {
+        return "failed to retrieve ODM error message";
+    } else {
+        return msg;
+    }
+}
+
+// ODM return value handling is kind of gross, so this saves us from
+// it a bit.  Even though it nominally returns a pointer, it can have
+// the value "-1" on error. In that case, we want to get the error
+// string and throw an exception. This handles all of that in all the
+// places we need it.
+#define check_odm_return(var, msg) if (reinterpret_cast<intptr_t>(var) == -1) { throw runtime_error((boost::format(msg ": %1%") % odm_error_string()).str()); }
+
+struct physical_processor
+{
+     string type;
+     long long frequency;
+     int smt_threads;
+     bool smt_enabled;
+};
+
+namespace facter { namespace facts { namespace aix {
+
+    processor_resolver::data processor_resolver::collect_data(collection& facts)
+    {
+        auto result = posix::processor_resolver::collect_data(facts);
+
+        // On AIX, we query the object data manager (odm) for
+        // processor information. This is a semi-hierarchical
+        // datastore of all the information about the system. For
+        // processors, we need to go through three links:
+        //
+        // 1. We query for all "predefined devices" with the
+        // "processor" class. I don't know if it's actually possible
+        // to get more than one result here - on the Puppet LPARs
+        // there is only one.
+        //
+        // 2. For each predefined device, we query the "custom
+        // devices". These represent the actual processors in the
+        // machine.
+        //
+        // 3. For each custom device, we query its attributes. These
+        // are things like frequency, type, and SMT information.
+
+        if (odm_initialize() < 0) {
+            throw runtime_error((boost::format("failed to initialize ODM: %1%") % odm_error_string()).str());
+        }
+        leatherman::util::scope_exit(odm_terminate);
```

Comment:
Good catch. Shouldn't this trip up some "result of expression not used" warning or lint somewhere? :(